src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -363,13 +387,25 @@ public NodeStateTable<T> ToImmutableAndFree()
                     return NodeStateTable<T>.Empty;
                 }
 
+#if DEBUG
+                // If the caller requested a specific capacity, then we should have added exactly that amount of items.
+                Debug.Assert(_requestedTableCapacity == null || _requestedTableCapacity == _states.Count);
+#endif
+
                 // if we added the exact same entries as before, then we can directly embed previous' entry array,
                 // avoiding a costly allocation of the same data.
-                var finalStates = _states.Count == _previous.Count && _states.SequenceEqual(_previous._states, (e1, e2) => e1.Matches(e2, _equalityComparer))
-                    ? _previous._states
-                    : _states.ToImmutable();
-
-                _states.Free();
```

Comment:
this is an interesting pitfall.  I originally wrote things this way so there was just a single `.Free` call.  However, that actually was bad here as `ToImmutable` cannot optimize by using `MoveToImmutable` under the covers (like `ToImmutableAndFree` does).